C# integration Docs (#2675)
saucelabs/sauce-docs: docs/visual-testing/integrations/csharp.md

---
sidebar_label: C#/.Net
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# C#/.Net WebDriver Integration

:::note Important
Access to this feature is currently limited to Enterprise customers as part of our commitment to providing tailored solutions. We are excited to announce that self-service access is under development and will be released shortly. Stay tuned!
:::

## Introduction

This guide requires an existing C# NUnit / xUnit setup.<br />
You can alternatively take a look to our [example repository](#examples).

Sauce Visual provides an library allowing integration with [WebDriver](https://www.selenium.dev/documentation/webdriver/).

> Sauce Visual plugin provides a library exposing a `VisualClient` object that provides actions:

Reviewer: 1.
```suggestion
Sauce Visual plugin provides a library exposing a `VisualClient` class that provides actions:
```

2. should we also mention about dispose?

> - `visualClient.VisualCheck()`: Takes a screenshot and send it to Sauce Visual for comparison.

Reviewer: ```suggestion
- `VisualCheck`: Takes a screenshot and sends it to Sauce Visual for comparison.
```

> - `visualClient.VisualResults()`: Waits for all diff calculations to complete and returns a summary of results.

Reviewer: ```suggestion
- `VisualResults`: Waits for all diff calculations to complete and returns a summary of results.
```

  See [Test results summary](#test-results-summary) for more details about summary format and sample usage.

## Quickstart

### Step 1: Add Sauce Visual dependecy

> Add [Sauce Visual](https://www.nuget.org/packages/SauceLabs.Visual/) dependency to your `.csproj`

Reviewer: normally you don't put such a code to show how to integrate - I recommend using nuget/dotnet reference instead:

```
Windows NuGet CLI:

`Install-Package SauceLabs.Visual`

Linux/Mac OS X .NET Core CLI:

`dotnet add package SauceLabs.Visual`
```

```xml
  <!-- FIXME USE Example Repo Reference -->
  <ItemGroup>
    ...
    <PackageReference Include="SauceLabs.Visual" Version="0.0.1" />
    ...
  </ItemGroup>
```

_Note: You can find the latest versions available [here](https://www.nuget.org/packages/SauceLabs.Visual#readme-body-tab)._

### Step 2: Configure Visual Testing integration

Declare a RemoteWebDriver and a VisualClient instance as class properties

```csharp
using OpenQA.Selenium.Remote;
using SauceLabs.Visual;

private RemoteWebDriver Driver { get; set; }
private VisualClient VisualClient { get; set; }
```

Initialize `RemoteWebDriver` and `VisualClient`

<Tabs
defaultValue="NUnit"
  values={[
    {label: 'NUnit', value: 'NUnit'},
    {label: 'xUnit', value: 'xUnit'},
  ]}>
<TabItem value="NUnit">

```csharp
    [OneTimeSetUp]
    public void Setup()
    {
      var sauceUsername = "YOUR_USERNAME";
      var sauceAccessKey = "YOUR_ACCESS_KEY";
      var sauceUrl = "https://ondemand.us-west-1.saucelabs.com:443/wd/hub";

      var browserOptions = new ChromeOptions();
      browserOptions.PlatformName = "Windows 11";
      browserOptions.BrowserVersion = "latest";

>       var sauceOptions = new Dictionary<string, object>();

Reviewer: ```suggestion
      var sauceOptions = new Dictionary<string, string>();
```

      sauceOptions.Add("username", sauceUsername);
      sauceOptions.Add("accessKey", sauceAccessKey);
      browserOptions.AddAdditionalOption("sauce:options", sauceOptions);

      Driver = new RemoteWebDriver(sauceUrl, browserOptions);

      VisualClient = new VisualClient(Driver, Region.UsWest1, sauceUsername, sauceAccessKey,
          new CreateBuildOptions() { Name = "My Visual Build", Project = "csharp-project", Branch = "csharp-branch" });

>       VisualClient.CaptureDom = true;

Reviewer: shouldn't we add a comment like 

// configure other options
VisualClient.CaptureDom = true

      sauceOptions.Add("username", sauceUsername);
      sauceOptions.Add("accessKey", sauceAccessKey);
      browserOptions.AddAdditionalOption("sauce:options", sauceOptions);

      Driver = new RemoteWebDriver(sauceUrl, browserOptions);

      VisualClient = new VisualClient(Driver, Region.UsWest1, sauceUsername, sauceAccessKey,
          new CreateBuildOptions() { Name = "My Visual Build", Project = "csharp-project", Branch = "csharp-branch" });
      VisualClient.CaptureDom = true;
    }
```

  </TabItem>
</Tabs>

To enhance efficiency in managing tests, it's important to provide a specific test name and suite name for each test. This practice allows Sauce Visual to effectively organize snapshots into coherent groups. As a result, it simplifies the review process, saving time and effort in navigating through test results and understanding the context of each snapshot.

Don't forget to quit the WebDriver and Dispose VisualClient.
<Tabs
defaultValue="NUnit"
  values={[
    {label: 'NUnit', value: 'NUnit'},
    {label: 'xUnit', value: 'xUnit'},
  ]}>
<TabItem value="NUnit">

```csharp
    [OneTimeTearDown]
    public void Teardown()
    {
        Driver?.Quit();

>         VisualClient.Cleanup().Wait();

Reviewer: We cannot suggest using `Wait()`

Author: That's something we can replace for NUnit, not sure about alternatives in xUnit, will dig deeper.

        VisualClient.Dispose();
    }
```

  </TabItem>
  <TabItem value="xUnit">

```csharp
  public void Dispose()
  {
      Driver?.Quit();
      VisualClient.Cleanup().Wait();
      VisualClient.Dispose();
  }
```

  </TabItem>
</Tabs>

### Step 3: Add visual tests in your tests

Add a check to one of your tests:

<Tabs
defaultValue="NUnit"
  values={[
    {label: 'NUnit', value: 'NUnit'},
    {label: 'xUnit', value: 'xUnit'},
  ]}>
<TabItem value="NUnit">

```csharp

    [Test]

>     public async Task SauceDemoHomePage()

Reviewer: ```suggestion
    public async Task SauceDemoHomePage_ShouldWaitForUserNameLocator_TheLocatorIsAvailableOnThePage()
```

WDYT?

Sauce Visual relies on environment variables for authentications.<br />
Both `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` need to be set prior starting your .Net job.

Username and Access Key can be retrieved from https://app.saucelabs.com/user-settings.

```sh
export SAUCE_USERNAME=__YOUR_SAUCE_USER_NAME__
export SAUCE_ACCESS_KEY=__YOUR_SAUCE_ACCESS_KEY__
```

### Step 5: Run the test

Upon executing your tests for the first time under this step, a visual baseline is automatically created in our system. This baseline serves as the standard for all subsequent WebDriver tests. As new tests are run, they are compared to this original baseline, with any deviations highlighted to signal visual changes. These comparisons are integral for detecting any unintended visual modifications early in your development cycle. All test builds, including the initial baseline and subsequent runs, can be monitored and managed through the Sauce Labs platform at [Sauce Visual Builds](https://app.saucelabs.com/visual/builds).

Remember, the baseline is established during the initial run, and any subsequent visual differences detected will be marked for review.

## Advanced usage

### Test results summary

> `VisualClient.VisualResults()` returns a summary of test results in `dictionnary<DiffStatus, int>` format where `DiffStatus` is one of the following:

Reviewer: ```suggestion
`VisualClient.VisualResults()` returns a summary of test results in `Dictionnary<DiffStatus, int>` format where `DiffStatus` is one of the following:
```

- `DiffStatus.QUEUED`: Diffs that are pending for processing. Should be 0 in case the test is completed without any timeouts
- `DiffStatus.EQUAL`: Diffs that have no changes detected
- `DiffStatus.UNAPPROVED`: Diffs that have detected changes and waiting for action
- `DiffStatus.APPROVED`: Diffs that have detected changes and have been approved
- `DiffStatus.REJECTED`: Diffs that have detected changes and have been rejected

Sample usage:

```csharp
var expectedTotalUnapprovedDiffs = 0;

var results = await VisualClient.VisualResults();

> Assert.AreEqual(expectedTotalUnapprovedDiffs, results?[DiffStatus.Unapproved]);

Reviewer: I think result is always defined, right?
```suggestion
Assert.AreEqual(expectedTotalUnapprovedDiffs, results[DiffStatus.Unapproved]);
```

---

> sidebar_label: C#/.Net

Reviewer 2: ```suggestion
sidebar_label: C#/DotNet
```

---

import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # C#/.Net WebDriver Integration

Reviewer 2: ```suggestion
# C#/DotNet WebDriver Integration
```

:::note Important
Access to this feature is currently limited to Enterprise customers as part of our commitment to providing tailored solutions. We are excited to announce that self-service access is under development and will be released shortly. Stay tuned!
:::

## Introduction

This guide requires an existing C# NUnit / xUnit setup.<br />
You can alternatively take a look to our [example repository](#examples).

Sauce Visual provides an library allowing integration with [WebDriver](https://www.selenium.dev/documentation/webdriver/).

Sauce Visual plugin provides a library exposing a `VisualClient` class that provides actions:

- `VisualCheck`: Takes a screenshot and sends it to Sauce Visual for comparison.
- `VisualResults`: Waits for all diff calculations to complete and returns a summary of results.
  See [Test results summary](#test-results-summary) for more details about summary format and sample usage.

## Quickstart

### Step 1: Add Sauce Visual dependecy

Add [Sauce Visual](https://www.nuget.org/packages/SauceLabs.Visual/) dependency to your `.csproj`

<Tabs
defaultValue="windows"
  values={[
    {label: 'Windows NuGet CLI', value: 'windows'},

>     {label: 'Linux/macOS/.Net Core', value: 'other'},

Reviewer 2: ```suggestion
    {label: 'Linux/macOS/DotNet Core', value: 'other'},
```

      var browserOptions = new ChromeOptions();
      browserOptions.PlatformName = "Windows 11";
      browserOptions.BrowserVersion = "latest";

      var sauceOptions = new Dictionary<string, string>();
      sauceOptions.Add("username", sauceUsername);
      sauceOptions.Add("accessKey", sauceAccessKey);
      browserOptions.AddAdditionalOption("sauce:options", sauceOptions);

      Driver = new RemoteWebDriver(sauceUrl, browserOptions);

      VisualClient = await VisualClient.Create(Driver, Region.UsWest1, new CreateBuildOptions()
      {
          Name = "My Visual Build",
          Project = "csharp-project",
          Branch = "csharp-branch"
      });
      // Enable Dom Capture
      VisualClient.CaptureDom = true;
    }
```

  </TabItem>
  <TabItem value="xUnit">

```csharp
  class MyTestClass : IAsyncLifetime {
    [...]
    public async Task InitializeAsync()
    {
        var browserOptions = Utils.GetBrowserOptions();
        var sauceOptions = Utils.GetSauceOptions();
        browserOptions.AddAdditionalOption("sauce:options", sauceOptions);

        var sauceUrl = Utils.GetOnDemandURL();
        Driver = new RemoteWebDriver(sauceUrl, browserOptions);

>         Driver.ExecuteScript("sauce:job-name=xUnit C#/.Net Visual Session");

Reviewer 2: ```suggestion
        Driver.ExecuteScript("sauce:job-name=xUnit C#/DotNet Visual Session");
```

    }
  }
```

  </TabItem>
</Tabs>

### Step 3: Add visual tests in your tests

Add a check to one of your tests:

<Tabs
defaultValue="NUnit"
  values={[
    {label: 'NUnit', value: 'NUnit'},
    {label: 'xUnit', value: 'xUnit'},
  ]}>
<TabItem value="NUnit">

```csharp

    [Test]
    public async Task SauceDemoHomePage_ShouldRenderLoginCorrectly()
    {
        Driver.Navigate().GoToUrl("https://www.saucedemo.com");
        var wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(15));
        wait.Until(drv => drv.FindElement(usernameLocator));

        await VisualClient.VisualCheck("Sauce Demo Homepage");
    }
```

  </TabItem>
  <TabItem value="xUnit">

```csharp

    [Fact]
    public async Task SauceDemoHomePage_ShouldRenderLoginCorrectly()
    {
        Driver.Navigate().GoToUrl("https://www.saucedemo.com");
        var wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(15));
        wait.Until(drv => drv.FindElement(usernameLocator));

        await VisualClient.VisualCheck("Sauce Demo Homepage");
    }
```

  </TabItem>
</Tabs>

### Step 4: Configure your Sauce Labs credentials

Sauce Visual relies on environment variables for authentications.<br />

> Both `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` need to be set prior starting your .Net job.

Reviewer 2: ```suggestion
Both `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` need to be set prior starting your DotNet job.
```
